pytensor/tensor/slinalg.py

```diff
@@ -791,8 +797,9 @@ def make_node(self, A, B):
     def perform(self, node, inputs, output_storage):
         (A, B) = inputs
         X = output_storage[0]
+        out_dtype = node.outputs[0].type.dtype
 
-        X[0] = scipy.linalg.solve_continuous_lyapunov(A, B)
+        X[0] = scipy.linalg.solve_continuous_lyapunov(A, B).astype(out_dtype)
```

Comment:
Why not let scipy natural output dtype? You would have to guess it. Is there undesirable upcast to float64 or something?

Reply:
If I remove it I get errors in the gradient tests:

```py
FAILED tests/tensor/test_slinalg.py::test_solve_discrete_are_grad[False] - TypeError: expected type_num 11 (NPY_FLOAT32) got 12

           TypeError: expected type_num 11 (NPY_FLOAT32) got 12
           Apply node that caused the error: Mul(random_projection, SolveDiscreteARE{enforce_Q_symmetric=True}.0)
           Toposort index: 1
           Inputs types: [TensorType(float64, shape=(None, None)), TensorType(float32, shape=(None, None))]
           Inputs shapes: [(2, 2), (2, 2)]
           Inputs strides: [(16, 8), (16, 8)]
           Inputs values: [array([[1.26416215, 1.05049824],
                  [1.04219109, 1.11393532]]), array([[14.5623059 ,  9.70820393],
                  [ 9.70820393,  6.47213595]])]
           Outputs clients: [[Sum{axes=None}(Mul.0)]]
```

I guess because `solve_discrete_are` calls `solve_discrete_lyapunov` in the gradient, which ends up getting upcast and causing the error

Reply:
The gradients should have the same type as the input so you may need a cast in the gradient? Anyway you're free to decide how the node behaves dtype wise.

Is this related to the problems you saw with pre-instantiating the Op once?

Reply:
I was having a problem with complex outputs being cast to float , but that seems to have gone away. 

This error is proving a bit more elusive. It's happening when the gradient test function is being compiled, but it's a numerical error? I guess it's evaluating the function to figure out dtypes?

Reply:
I need some guidance on how to handle the scipy upcasting. 

If I leave everything as-is, we get unexpected type errors because the dtype coming out of the `perform` method doesn't match the node's expected output type, but only inside the `vm` after compiling a function. Example:

```
import pytensor
import pytensor.tensor as pt
import numpy as np

a, b, q, r = (
    np.array([[4, 3], [-4.5, -3.5]]),
    np.array([[1], [-1]]),
    np.array([[9, 6], [6, 4]]),
    np.array([[1]]),
)
a, b, q, r = (pt.as_tensor_variable(x).astype('float32') for x in [a, b, q, r])

x = pt.linalg.solve_discrete_are(a, b, q, r)
print(x.type.dtype) # float32
print(x.eval()) # works

y = pt.tensor('y', dtype='float32', shape=(2,2))
z = (x * y).sum()
print(z.eval({y:np.eye(2).astype('float32')}).dtype) # float32, works

f = pytensor.function([a,b,q,r,y], z)

# Raises
f(*[var.eval() for var in [a, b, q, r]], np.eye(2).astype('float32'))

```

Reply:
So first thing first, what does `x.eval().type` show? 

Taking a step back. The type returned by `make_node` and the one evaluated in `perform` must match. This is the promise that `make_node` is doing > Here is an Apply node that takes these types as inputs and returns these output types. Not respecting this can cause errors like you found, although sometimes it goes unseen, as when it is the direct output of a function. We don't try and check the promise is kept by every perform call, as that could be pretty costly.

You are free to make `make_node` and `perform` agree to each other however you want. In most cases you would want to respect (and therefore predict) what scipy does naturally to avoid useless casting. Sometimes it's hard to predict the type, so some Ops do a lazy check: call the perform function with the input types to figure out the output dtype. Example:   https://github.com/pymc-devs/pytensor/blob/a377c22dd075fda3b1ef0c0fa134bce01cfe7af3/pytensor/tensor/nlinalg.py#L314-L325

> If I do that though, I hit the complex dtype bug, where if I first run a float test then a complex test, the dtype of node.outputs[0].type.dtype for the BilinearSolveDiscreteLyapunov Op is not updated from float to complex. All tests pass if I run them one-by-one, however.

This I don't understand. Are you saying somehow `make_node` is not being called the second time (perhaps a clone_replace was done, and it didn't figure out it should rebuild the node)? You can dprint the calls to make_node in your Op (and the input-output types) to see it.

Reply:
This thread was resolved by #1048 , thanks for a quick fix.